File: src/main/scala/Location.scala
object Location {
  def beginningOfFile(fileName: String) = Location(fileName, 0, 0, 0, 0)
  def at(fileName: String, line: Int, column: Int) = Location(fileName, line, column, line, column)
}

case class Location(
  fileName: String,
  startLine: Int,
  startCol: Int,
  endLine: Int,
  endCol: Int
) {
  def extendWith(other: Location): Location = {
    if (fileName != other.fileName) {
      throw new IllegalArgumentException("Locations are not of the same file")
    }

    copy(endLine = other.endLine, endCol = other.endCol)
  }

  override def toString: String = {
    if (startLine == endLine && startCol == endCol)
      s"$fileName:$startLine:$startCol"
    else
      s"$fileName:(from $startLine:$startCol to $endLine:$endCol)"
  }
}

File: src/main/kotlin/photon/core/Location.kt
package photon.core

data class Location(
  val fileName: String,
  val startLine: Int,
  val startCol: Int,
  val endLine: Int,
  val endCol: Int
) {
  companion object {
    fun beginningOfFile(fileName: String): Location {
      return Location(fileName, 0, 0, 0, 0)
    }

    fun at(fileName: String, line: Int, column: Int): Location {
      return Location(fileName, line, column, line, column)
    }
  }

  fun extendWith(other: Location): Location {
    if (fileName != other.fileName) {
      throw IllegalArgumentException("Locations are not of the same file")
    }

    return copy(endLine = other.endLine, endCol = other.endCol)
  }

  override fun toString(): String {
    if (startLine == endLine && startCol == endCol)
      return "$fileName:$startLine:$startCol"
    else
      return "$fileName:(from $startLine:$startCol to $endLine:$endCol)"
  }
}